Allow lockfile path to be configured

dotboris opened the pull request:
The goal of this feature is to allow users to control the path of the lockfile. This is useful in situations where you don't want the lockfile to default next to the config file.

For my use case, I want to generate the config file with Nix which will store it in a read-only path. In such a situation `autorestic` won't be able to create the lockfile.

This PR adds a few ways to let users control their lockfile path:
- Using the new `--lockfile` flag
- Using the new `lockfile: ...` config option

Note that this PR includes a refactor that merges the `lock` package back into the `internal` package. This is done to avoid a circular dependency where the `lock` package needs to read the config from the `internal` package but the `internal` package also needs to use the `lock` package.

dotboris committed:
feat(lockfile): allow changing the lockfile path

dotboris committed:
doc(lockfile): Document `--lockfile-path` flag

PHLAK commented:
I just started using autorestic and was looking for a way to control this. My reasoning is that I'm putting the config file in `/etc/autorestic` and running it on a schedule with a systemd timer but having the lock file in `/etc` doesn't make sense. I want to put it in `/var/run/autorestic` or somewhere similar.

PHLAK commented:
It would be nice if we could set the lockfile path in the config file similar to the `cache-dir`.

```yaml
version: 2

global:
  all:
    cache-dir: /var/cache/restic
    lockfile-dir: /run/lock/restic/
```

dotboris commented:
I like the idea of specifying the lockfile through the config file. It makes a ton of sense to me.

I don't think that we can fit this under the `global` section tho. That section is meant to host options (as in command line options) that are passed to `restic` based on the command line. In your example, `global.all.cache-dir = /var/cache/restic` makes it so that every restic command gets called with `--cache-dir /var/cache/restic`. Since this lockfile is an `autorestic` concept it can't really live in there.

What do you think of adding this option at the root of the config. So it'd be something like this:

```yaml
version: 2
lockfile: /var/lock/restic/autorestic.lock.yaml

global:
  all:
    cache-dir: ...

...
```

Note that in my example `lockfile` points to a file and not a directory because there's really only one lockfile for `autorestic`.

PHLAK commented:
> I don't think that we can fit this under the `global` section tho. That section is meant to host options (as in command line options) that are passed to `restic` based on the command line.

That's true. Making it a top-level configuration option makes sense then.

> Note that in my example `lockfile` points to a file and not a directory because there's really only one lockfile for `autorestic`.

I was debating this myself in my example so I think this also makes sense.

dotboris committed:
Merge branch 'master' into lockfile-path-config

dotboris committed:
chore: move lockfile code to internal module

In a future commit, I'll need the lockfile code to access the config
file. This solves an import cycle.

dotboris committed:
feat(config): allow specifying lockfile

dotboris commented:
I've implemented the basic functionality of reading the lockfile path from the config. I still need to do some of the finishing touches: documentation and tests.

dotboris committed:
chore(lock): test setting lockfile path from config

dotboris committed:
doc(lock): add lockfile doc with customization explanation

dotboris committed:
chore(lockfile) change flag to --lockfile

This is more consistent with the `lockfile` config option

dotboris commented:
I just finished up. The documentation and tests are now done. On top of this, I've changed the flag from `--lockfile-path` to `--lockfile` so that it matches the new `lockfile: ...` config option.

dcelasun commented:
@dotboris mind squashing commits and rebasing this from master?

@cupcakearmy I think this is good to go.

dotboris committed:
Merge branch 'master' into lockfile-path-config

dotboris commented:
:wave: I've stopped using `autorestic` and I therefore don't really have a need for this PR anymore. As such, I won't be maintaining this PR or keeping it open. If someone wants to take over my work and get this over the finish line feel free to do so.
